diff --git a/src/useractions.cpp b/src/useractions.cpp
index a77a8ace18..4af8ebb28f 100644
--- a/src/useractions.cpp
+++ b/src/useractions.cpp
@@ -210,7 +210,7 @@
     connect(m_menu, &QMenu::aboutToHide, this, &UserActionsMenu::menuAboutToHide, Qt::QueuedConnection);
     connect(m_menu, &QMenu::triggered, this, &UserActionsMenu::slotWindowOperation);
 
-    QMenu *advancedMenu = new QMenu(m_menu);
+    QMenu *advancedMenu = m_menu;
     connect(advancedMenu, &QMenu::aboutToShow, this, [this, advancedMenu]() {
         if (m_window) {
             advancedMenu->setPalette(m_window->palette());
@@ -287,10 +287,6 @@
     m_keepAboveOperation->setCheckable(true);
     m_keepAboveOperation->setData(Options::KeepAboveOp);
 
-    QAction *overflowAction = m_menu->addMenu(advancedMenu);
-    overflowAction->setText(i18n("&More Actions"));
-    overflowAction->setIcon(QIcon::fromTheme(QStringLiteral("overflow-menu")));
-
     m_menu->addSeparator();
 
     m_closeOperation = m_menu->addAction(i18n("&Close"));
