diff --git a/lix/libstore/build/local-derivation-goal.cc b/lix/libstore/build/local-derivation-goal.cc
index 57b679ae58..aacadf6d6f 100644
--- a/lix/libstore/build/local-derivation-goal.cc
+++ b/lix/libstore/build/local-derivation-goal.cc
@@ -965,6 +965,7 @@
     /* Fork a child to build the package. */
     this->pg = ProcessGroup{startChild(std::move(setupChildFD), std::move(builderOut))};
     auto & pg = std::get<ProcessGroup>(this->pg);
+    printWarning("started progress group %d for '%s'", pg.leader.get(), drv->name);
 
     // set up the child. we do not use rpc here because we want the build launcher to be
     // as simple as possible, and that excludes using an event loop. we'll handle errors
diff --git a/lix/libstore/daemon.cc b/lix/libstore/daemon.cc
index e129f552e2..ec3a04ec47 100644
--- a/lix/libstore/daemon.cc
+++ b/lix/libstore/daemon.cc
@@ -90,6 +90,11 @@
             return BufferState::HasSpace;
         }
 
+        if (lvl <= lvlNotice) {
+            // For warning-level or higher, log both locally and remotely.
+            writeLogsToStderr(s + "\n");
+        }
+
         StringSink buf;
         buf << STDERR_NEXT << (s + "\n");
         return enqueueMsg(buf.s);
diff --git a/lix/libstore/platform/linux.cc b/lix/libstore/platform/linux.cc
index 4bbd08021e..2a99a08ec8 100644
--- a/lix/libstore/platform/linux.cc
+++ b/lix/libstore/platform/linux.cc
@@ -36,6 +36,7 @@
 #include <netinet/in.h>
 #include <numeric>
 #include <optional>
+#include <print>
 #include <regex>
 #include <sched.h>
 #include <string_view>
@@ -1203,7 +1204,12 @@
             buildUser->getGID()
         );
 
-        debug("using cgroup '%s' for build", context.cgroup->name());
+        notice("using cgroup '%s' for build", context.cgroup->name());
+        std::println(std::cerr, "using cgroup '{}' for build '{}'\n", context.cgroup->name(), drvPath.name());
+        //if (this->env.contains("name")) {
+        //    std::println(std::cerr, "using cgroup '{}' for build '{}'\n", context.cgroup->name(), this->env["name"]);
+        //} else {
+        //}
 
         /* TODO(raito): it would be very nice if we could propagate system features
          * based on which cgroup controllers are available in `context.cgroup`
diff --git a/lix/libutil/async-io.cc b/lix/libutil/async-io.cc
index d6ff1d0e3f..89a85fd22c 100644
--- a/lix/libutil/async-io.cc
+++ b/lix/libutil/async-io.cc
@@ -194,7 +194,7 @@
     } else if (errno == EAGAIN || errno == EWOULDBLOCK) {
         return observer.whenBecomesWritable().then([=, this] { return write(src, size); });
     } else {
-        return {result::failure(std::make_exception_ptr(SysError(errno, "write failed")))};
+        return {result::failure(std::make_exception_ptr(SysError(errno, "write failed for %d", fd)))};
     }
 }
 AsyncFramedInputStream::~AsyncFramedInputStream()
diff --git a/lix/libutil/async-io.hh b/lix/libutil/async-io.hh
index 064231e9cb..83d589b9f8 100644
--- a/lix/libutil/async-io.hh
+++ b/lix/libutil/async-io.hh
@@ -198,6 +198,7 @@
     FdBlockingState oldState;
     AutoCloseFD ownedFd; // only for closing automatically, must equal fd if set
     kj::UnixEventPort::FdObserver observer;
+    std::optional<std::string> label = std::nullopt;
 
 public:
     struct shared_fd
diff --git a/lix/libutil/processes.hh b/lix/libutil/processes.hh
index 9d44f93d76..a33026e17e 100644
--- a/lix/libutil/processes.hh
+++ b/lix/libutil/processes.hh
@@ -41,9 +41,9 @@
 
 class ProcessGroup
 {
+public:
     Pid leader;
 
-public:
     ProcessGroup() = default;
     explicit ProcessGroup(Pid leader) : leader(std::move(leader)) {}
     ProcessGroup(ProcessGroup &&) = default;
